lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +42,57 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(config: PbenchServerConfig, logger: Logger):
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server need to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+    """
+    try:
+        oidc_server = config.get(
+            "authentication",
+            "internal_server_url",
+            fallback=config.get("authentication", "server_url"),
+        )
+    except (NoSectionError):
+        logger.exception("Bad config file:  missing 'authentication' section")
+        sys.exit(1)
+    except (NoOptionError):
+        logger.error("Bad config file:  no 'server_url' in  'authentication' section")
+        sys.exit(1)
+
+    session = requests.Session()
+    # The connection check will retry thrice unless successful, viz.,
+    # [0.0s, 20.0s, 40.0s]. urllib3 will sleep for:
+    # {backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between
+    # the retries. More detailed documentation on Retry and backoff_factor can
+    # be found at:
+    # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
+    retry = Retry(total=3, backoff_factor=10)
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+    response = session.get(f"{oidc_server}/health")
```

Comment:
Are you sure that you want to use a `requests.Session` for this connection?  It seems like it would be simpler to just use `http.request()` and specify the retry object (as in the middle example [here](https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#urllib3.util.Retry)).  The object returned would be a [`urllib3.response.HTTPResponse`](https://urllib3.readthedocs.io/en/latest/reference/urllib3.response.html#urllib3.response.HTTPResponse) (not a `requests.Response`), but it has a `status` attribute and a `json()` method, so it shouldn't be hard to adapt.

Reply:
Yeah, I pondered upon that for a little while but decide to use `Session` object thinking we can reuse this session for checking other connections like `postgresql` or `elasticsearch` if we decide to do more active connection checks in our shell. But I guess `http.request()` would also be okay for our use case.

Reply:
That's an excellent point -- if we going to be making a series of requests, then the session is probably worthwhile.  (I was just thinking that it was overkill for a single request.)  And, since we're accustomed to using the `requests` package, not confusing things by using `urllib3.response` would probably be a good thing.  (Deep down, I bet they are all using the same implementation, but I don't really know....)

I mainly wanted to offer you the alternative if it seemed attractive.